README.md
## NTPClient_Generic Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/NTPClient_Generic.svg?)](https://www.ardu-badge.com/NTPClient_Generic)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/NTPClient_Generic.svg)](https://github.com/khoih-prog/NTPClient_Generic/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/NTPClient_Generic/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/NTPClient_Generic.svg)](http://github.com/khoih-prog/NTPClient_Generic/issues)

<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---

## Table of Contents

* [Important Change from v3.7.0](#Important-Change-from-v370)
* [Why do we need this NTPClient_Generic library](#why-do-we-need-this-ntpclient_generic-library)
  * [Features](#features)
  * [Currently supported Boards](#currently-supported-boards)
  * [Currently supported WiFi shields/modules](#currently-supported-wifi-shieldsmodules)
  * [Currently supported Ethernet shields/modules](#currently-supported-ethernet-shieldsmodules)
* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [Packages' Patches](#packages-patches)
  * [1. For Adafruit nRF52840 and nRF52832 boards](#1-for-adafruit-nRF52840-and-nRF52832-boards)
  * [2. For Teensy boards](#2-for-teensy-boards)
  * [3. For Arduino SAM DUE boards](#3-for-arduino-sam-due-boards)
  * [4. For Arduino SAMD boards](#4-for-arduino-samd-boards)
      * [For core version v1.8.10+](#for-core-version-v1810)
      * [For core version v1.8.9-](#for-core-version-v189-)
  * [5. For Adafruit SAMD boards](#5-for-adafruit-samd-boards)
  * [6. For Seeeduino SAMD boards](#6-for-seeeduino-samd-boards)
  * [7. For STM32 boards](#7-for-stm32-boards) 
    * [7.1. For STM32 boards to use LAN8720](#71-for-stm32-boards-to-use-lan8720)
    * [7.2. For STM32 boards to use Serial1](#72-for-stm32-boards-to-use-serial1)
  * [8. For RP2040-based boards using Earle Philhower arduino-pico core](#8-for-rp2040-based-boards-using-earle-philhower-arduino-pico-core)
    * [8.1. To use BOARD_NAME](#81-to-use-board_name)
    * [8.2. To avoid compile error relating to microsecondsToClockCycles](#82-to-avoid-compile-error-relating-to-microsecondstoclockcycles)
  * [9. For Portenta_H7 boards using Arduino IDE in Linux](#9-for-portenta_h7-boards-using-arduino-ide-in-linux)
  * [10. For RTL8720DN boards using AmebaD core](#10-for-rtl8720dn-boards-using-amebad-core)
  * [11. For SAMD21 and SAMD51 boards using ArduinoCore-fab-sam core](#11-For-SAMD21-and-SAMD51-boards-using-ArduinoCore-fab-sam-core)
  * [12. For Seeeduino RP2040 boards](#12-For-Seeeduino-RP2040-boards)
* [Libraries' Patches](#libraries-patches)
  * [1. For application requiring 2K+ HTML page](#1-for-application-requiring-2k-html-page)
  * [2. For Ethernet library](#2-for-ethernet-library)
  * [3. For EthernetLarge library](#3-for-ethernetlarge-library)
  * [4. For Etherne2 library](#4-for-ethernet2-library)
  * [5. For Ethernet3 library](#5-for-ethernet3-library)
  * [6. For UIPEthernet library](#6-for-uipethernet-library)
  * [7. For fixing ESP32 compile error](#7-for-fixing-esp32-compile-error)
  * [8. For fixing ESP8266 compile error](#8-for-fixing-esp8266-compile-error)
* [HOWTO Fix `Multiple Definitions` Linker Error](#howto-fix-multiple-definitions-linker-error)
* [HOWTO Use analogRead() with ESP32 running WiFi and/or BlueTooth (BT/BLE)](#howto-use-analogread-with-esp32-running-wifi-andor-bluetooth-btble)
  * [1. ESP32 has 2 ADCs, named ADC1 and ADC2](#1--esp32-has-2-adcs-named-adc1-and-adc2)
  * [2. ESP32 ADCs functions](#2-esp32-adcs-functions)
  * [3. ESP32 WiFi uses ADC2 for WiFi functions](#3-esp32-wifi-uses-adc2-for-wifi-functions)
* [Important Notes](#important-notes)
* [Examples](#examples)
  * [ESP32/ESP8266 Boards](#esp32esp8266-boards)
  * [Generic Boards with Ethernet](#generic-boards-with-ethernet)
  * [STM32F/L/H/G/WB/MP1 Boards with Ethernet](#stm32flhgwbmp1-boards-with-ethernet)
  * [Generic Boards with WiFiNINA](#generic-boards-with-wifinina)
  * [WT32_ETH01 Boards](#wt32_eth01-boards)
  * [RTL8720DN Boards](#rtl8720dn-boards)
  * [Portenta_H7 Boards with WiFi](#Portenta_H7-Boards-with-WiFi)
  * [Portenta_H7 Boards with Ethernet](#Portenta_H7-Boards-with-Ethernet)
  * [General](#General)
* [Example RTC_Ethernet_NTPClient](#example-rtc_ethernet_ntpclient)
  * [ 1. File RTC_Ethernet_NTPClient.ino](#1-file-rtc_ethernet_ntpclientino)
  * [ 2. File defines.h](#2-file-definesh)
* [Debug Terminal Output Samples](#debug-terminal-output-samples)
  * [ 1. Ethernet_NTPClient_Advanced on Adafruit NRF52840_FEATHER_EXPRESS with ENC28J60 using EthernetENC Library](#1-ethernet_ntpclient_advanced-on-adafruit-nrf52840_feather_express-with-enc28J60-using-ethernetenc-library)
  * [ 2. NINA_NTPClient_Advanced on Arduino SAMD21 SAMD_NANO_33_IOT with WiFiNINA using WiFiNINA_Generic Library](#2-nina_ntpclient_advanced-on-arduino-samd21-samd_nano_33_iot-with-wifinina-using-wifinina_generic-library)
  * [ 3. TZ_NTP_WorldClock_Ethernet on Adafruit NRF52840_FEATHER_EXPRESS with W5500 using Ethernet_Generic Library](#3-tz_ntp_worldclock_ethernet-on-adafruit-nrf52840_feather_express-with-w5500-using-Ethernet_Generic-library)
  * [ 4. TZ_NTP_WorldClock_STM32_Ethernet on STM32F7 Nucleo-144 NUCLEO_F767ZI with LAN8742A using STM32Ethernet Library](#4-tz_ntp_worldclock_stm32_ethernet-on-stm32f7-nucleo-144-nucleo_f767zi-with-lan8742a-using-stm32ethernet-library)
  * [ 5. RTC_Ethernet_NTPClient on Adafruit NRF52840_FEATHER_EXPRESS with ENC28J60 using EthernetENC Library](#5-rtc_ethernet_ntpclient-on-adafruit-nrf52840_feather_express-with-enc28J60-using-ethernetenc-library)
  * [ 6. RTC_Ethernet_NTPClient_STM32 on STM32F7 Nucleo-144 NUCLEO_F767ZI with LAN8742A using STM32Ethernet Library](#6-rtc_ethernet_ntpclient_stm32-on-stm32f7-nucleo-144-nucleo_f767zi-with-lan8742a-using-stm32ethernet-library)
  * [ 7. ESP_NTPClient_Advanced on ESP8266_NODEMCU_ESP12E](#7-esp_ntpclient_advanced-on-esp8266_nodemcu_esp12e)
  * [ 8. ESP_NTPClient_Advanced on ESP32_DEV](#8-esp_ntpclient_advanced-on-esp32_dev)
  * [ 9. Ethernet_NTPClient_Advanced on SAM DUE with W5100 using Ethernet_Generic Library](#9-ethernet_ntpclient_advanced-on-sam-due-with-w5100-using-Ethernet_Generic-library)
  * [10. BI_RTC_Ethernet_NTPClient_STM32 on STM32F7 Nucleo-144 NUCLEO_F767ZI with LAN8742A using STM32Ethernet Library](#10-bi_rtc_ethernet_ntpclient_stm32-on-stm32f7-nucleo-144-nucleo_f767zi-with-lan8742a-using-stm32ethernet-library)
  * [11. BI_RTC_Ethernet_NTPClient_STM32 on STM32F7 Nucleo-144 NUCLEO_F767ZI with ENC28J60 using EthernetENC Library](#11-bi_rtc_ethernet_ntpclient_stm32-on-stm32f7-nucleo-144-nucleo_f767zi-with-enc28J60-using-ethernetenc-library)
  * [12. BI_RTC_Alarm_Ethernet_NTPClient_STM32 on STM32F7 Nucleo-144 NUCLEO_F767ZI with LAN8742A using STM32Ethernet Library](#12-bi_rtc_alarm_ethernet_ntpclient_stm32-on-stm32f7-nucleo-144-nucleo_f767zi-with-lan8742a-using-stm32ethernet-library)
  * [13. NINA_NTPClient_Basic on MBED NANO_RP2040_CONNECT with WiFiNINA using WiFiNINA_Generic Library](#13-nina_ntpclient_basic-on-mbed-nano_rp2040_connect-with-wifinina-using-wifinina_generic-library)
  * [14. Ethernet_NTPClient_Advanced on RASPBERRY_PI_PICO with W5x00 using Ethernet_Generic Library](#14-ethernet_ntpclient_advanced-on-raspberry_pi_pico-with-w5x00-using-Ethernet_Generic-library)
  * [15. Ethernet_NTPClient_Advanced on MBED RASPBERRY_PI_PICO with W5x00 using Ethernet_Generic Library](#15-ethernet_ntpclient_advanced-on-mbed-raspberry_pi_pico-with-w5x00-using-Ethernet_Generic-library)
  * [16. NINA_TZ_NTP_WorldClock on MBED NANO_RP2040_CONNECT with WiFiNINA using WiFiNINA_Generic Library](#16-nina_tz_ntp_worldclock-on-mbed-nano_rp2040_connect-with-wifinina-using-wifinina_generic-library)
  * [17. WT32_ETH01_TZ_NTP_WorldClock on ESP32_DEV with ETH_PHY_LAN8720](#17-wt32_eth01_tz_ntp_worldclock-on-esp32_dev-with-eth_phy_lan8720)
  * [18. WT32_ETH01_NTPClient_Advanced on ESP32_DEV with ETH_PHY_LAN8720](#18-wt32_eth01_ntpclient_advanced-on-esp32_dev-with-eth_phy_lan8720)
  * [19. RTL8720DN_TZ_NTP_WorldClock on Rtlduino RTL8720DN](#19-rtl8720dn_tz_ntp_worldclock-on-rtlduino-rtl8720dn)
  * [20. Portenta_H7_NTPClient_Advanced on PORTENTA_H7_M7 with WiFi](#20-Portenta_H7_NTPClient_Advanced-on-PORTENTA_H7_M7-with-WiFi)
  * [21. Portenta_H7_NTPClient_Advanced on PORTENTA_H7_M7 with Ethernet](#21-Portenta_H7_NTPClient_Advanced-on-PORTENTA_H7_M7-with-Ethernet)
  * [22. ESP_WiFi_ETH_NTPClient_Advanced on ESP32_DEV with W5x00 using Ethernet_Generic Library](#22-ESP_WiFi_ETH_NTPClient_Advanced-on-ESP32_DEV-with-W5x00-using-Ethernet_Generic-Library)
* [Debug](#debug)
* [Troubleshooting](#troubleshooting)
* [TO DO](#to-do)
* [DONE](#done)
* [Contributions and Thanks](#contributions-and-thanks)
* [Contributing](#contributing)
* [License](#license)
* [Copyright](#copyright)

---
---

### Important Change from v3.7.0

Please have a look at [HOWTO Fix `Multiple Definitions` Linker Error](#howto-fix-multiple-definitions-linker-error)

From v3.7.2, you can use the new `setUDP()` function for auto-switching between `WiFi` and `Ethernet` UDP to update NTP time.

Check examples [ESP_WiFi_ETH_NTPClient_Advanced](https://github.com/khoih-prog/NTPClient_Generic/tree/main/examples/ESP/ESP_WiFi_ETH_NTPClient_Advanced) and [WT32_WiFi_ETH01_NTPClient_Advanced](https://github.com/khoih-prog/NTPClient_Generic/tree/main/examples/WT32_ETH01/WT32_WiFi_ETH01_NTPClient_Advanced) for a demo.

---
---

### Why do we need this [NTPClient_Generic library](https://github.com/khoih-prog/NTPClient_Generic)

#### Features

The [**NTPClient_Generic library**](https://github.com/khoih-prog/NTPClient_Generic) is designed to work in conjunction with the [**Arduino Time library**](https://github.com/PaulStoffregen/Time), which must also be installed on your system. This documentation assumes some familiarity with the Time library.

The primary goal of the [**NTPClient_Generic library**](https://github.com/khoih-prog/NTPClient_Generic) is to enable the creation of an NTPClient to connect to NTP time-servers to get time and keep local time in sync. For **ESP8266/ESP32, WT32_ETH01 (ESP32 + LAN8720), nRF52, SAMD21/SAMD51, STM32F/L/H/G/WB/MP1, Teensy, RP2040-based boards (Nano_RP2040_Connect, RASPBERRY_PI_PICO), etc. besides AVR, using ESP WiFi, WiFiNINA, Ethernet W5x00, ENC28J60, LAN8742A, ESP8266/ESP32 AT-command WiFi**. But whether hardware RTC or another time-keeping source is present is not mandatory, since the [**Time library**](https://github.com/PaulStoffregen/Time) library can function as a software RTC without additional hardware (although its accuracy is dependent on the accuracy of the microcontroller's system clock.)

The examples will demonstrate how to get the UTC time from NTP server, then update the DS3231 RTC or local clock to make sure the time is perfectly correct.

In the examples, the interval to get perfect time from the NTP server to update RTC and/or local clock is configurable to ensure the RTC and/or local clock accuracy.

This [**NTPClient_Generic library**](https://github.com/khoih-prog/NTPClient_Generic) is based on and modified from [**Fabrice Weinberg's NTPClient Library**](https://github.com/arduino-libraries/NTPClient) to add functions, features and as well as support to many boards and shields.

---

#### Currently supported Boards

This [**NTPClient_Generic** library](https://github.com/khoih-prog/NTPClient_Generic) currently supports these following boards:

 1. **nRF52 boards**, such as **AdaFruit Feather nRF52832, nRF52840 Express, BlueFruit Sense, Itsy-Bitsy nRF52840 Express, Metro nRF52840 Express, NINA_B302_ublox, NINA_B112_ublox, etc.**
 
 2. **SAMD21**

changelog.md
## NTPClient_Generic Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/NTPClient_Generic.svg?)](https://www.ardu-badge.com/NTPClient_Generic)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/NTPClient_Generic.svg)](https://github.com/khoih-prog/NTPClient_Generic/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/NTPClient_Generic/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/NTPClient_Generic.svg)](http://github.com/khoih-prog/NTPClient_Generic/issues)

---
---

## Table of Contents

* [Changelog](#changelog)
  * [Releases v3.7.5](#releases-v375)
  * [Releases v3.7.4](#releases-v374)
  * [Releases v3.7.3](#releases-v373)
  * [Releases v3.7.2](#releases-v372)
  * [Releases v3.7.1](#releases-v371)
  * [Releases v3.7.0](#releases-v370)
  * [Releases v3.6.0](#releases-v360)
  * [Releases v3.5.2](#releases-v352)
  * [Releases v3.5.1](#releases-v351)
  * [Releases v3.5.0](#releases-v350)
  * [Releases v3.4.1](#releases-v341)
  * [Releases v3.4.0](#releases-v340)
  * [Releases v3.3.0](#releases-v330)
  * [Releases v3.2.2](#releases-v322)
  * [Releases v3.2.1](#releases-v321)

---
---

## Changelog

### Releases v3.7.5

1. Fix bug causing time jumping back or forth when network has problem. Check [Time jumping back when connection is lost and forward when up #9](https://github.com/khoih-prog/NTPClient_Generic/issues/9) and [Time jumping back 9 minutes #1](https://github.com/khoih-prog/NTPClient_Generic/issues/1)
2. Modify example [ESP_NTPClient_Advanced](https://github.com/khoih-prog/NTPClient_Generic/tree/main/examples/ESP/ESP_NTPClient_Advanced) to demo the better way to use the library, without using long `delay()`
3. Add support to `RP2040W` boards using CYW43439 WiFi
4. Optimize and clean up
5. Modify `platformio.ini` and `library.json` to reduce error with PIO
6. Update `Packages' Patches`

### Releases v3.7.4

1. Sync with [NTPClient releases v3.2.1](https://github.com/arduino-libraries/NTPClient/releases/tag/3.2.1)


### Releases v3.7.3

1. Use new [Ethernet_Generic library](https://github.com/khoih-prog/Ethernet_Generic) as default for W5x00.
2. Rewrite to avoid `multiple-definitions` linker error for multiple-file project
3. Add example [multiFileProject](examples/multiFileProject) to demo how to avoid `multiple-definitions` linker error for multiple-file project
4. Add support to SPI1 for RP2040 using [arduino-pico core](https://github.com/earlephilhower/arduino-pico)
5. Add support SPI2 for ESP32
6. Rewrite all the examples to support new features
7. Update `Packages' Patches`

### Releases v3.7.2

1. Add setUDP() function to enable auto-switching between `WiFi` and `Ethernet` UDP instances. Check [added setUDP function #8](https://github.com/khoih-prog/NTPClient_Generic/pull/8)
2. Add new examples [ESP_WiFi_ETH_NTPClient_Advanced](https://github.com/khoih-prog/NTPClient_Generic/tree/main/examples/ESP/ESP_WiFi_ETH_NTPClient_Advanced) and [WT32_WiFi_ETH01_NTPClient_Advanced](https://github.com/khoih-prog/NTPClient_Generic/tree/main/examples/WT32_ETH01/WT32_WiFi_ETH01_NTPClient_Advanced) to use the new setUDP() function for auto-switching between `WiFi` and `Ethernet` UDP to update NTP time.
3. Update `Packages_Patches`

### Releases v3.7.1

1. Fix getUTCEpochMillis() bug. Check [getUTCEpochMillis() not returning correct value #7](https://github.com/khoih-prog/NTPClient_Generic/issues/7)
2. Update Adafruit SAMD `Packages_Patches`

### Releases v3.7.0

1. Make compatible with old code
2. Update all examples
3. Update `Packages_Patches`

### Releases v3.6.0

1. Add support to Portenta_H7 using Murata WiFi of Vision-Shield Ethernet
2. Fix `multiple-definitions` linker error. Check [Different behaviour using the src_cpp or src_h lib #80](https://github.com/khoih-prog/ESPAsync_WiFiManager/discussions/80)
3. Update all examples
4. Update `Packages_Patches`

### Releases v3.5.2

1. Auto detect ESP32 core version.
2. Fix bug in examples for WT32_ETH01
3. Update `Packages' Patches`

### Releases v3.5.1

1. Update `platform.ini` and `library.json` to use original `khoih-prog` instead of `khoih.prog` after PIO fix
2. Update `Packages' Patches`

### Releases v3.5.0

1. Add support to RTL8720DN, RTL8722DM, RTL8722CSM, etc. boards
2. Add examples with new features

### Releases v3.4.1

1. Fix typo in `library.json`

### Releases v3.4.0

1. Add support to WT32_ETH01 (ESP32 + LAN8720) boards
2. Add examples with new features

### Releases v3.3.0

1. Add support to RP2040-based boards, such as **RASPBERRY_PI_PICO, ADAFRUIT_FEATHER_RP2040 and GENERIC_RP2040**, using [**Earle Philhower's arduino-pico** core](https://github.com/earlephilhower/arduino-pico).
2. Add support to RP2040-based boards, such as **Nano_RP2040_Connect, RASPBERRY_PI_PICO, ADAFRUIT_FEATHER_RP2040 and GENERIC_RP2040**, using [**Arduino-mbed RP2040** core](https://github.com/arduino/ArduinoCore-mbed).
3. Add packet validity checks
4. Add version string
5. Improve and add more debugging codes

### Releases v3.2.2

1. Add examples to use **STM32 Built-In RTC.**

### Releases v3.2.1

1. Initial porting to many Generic boards using WiFi/Ethernet modules/shields.
2. Add support to **SAMD21/SAMD51, nRF52, STM32F/L/H/G/WB/MP1 with WiFiNINA, W5x00, ENC28J60, LAN8742A Ethernet modules/shields.**
3. Add functions.
4. Completely new examples using NTP time to update DS3231 RTC and/or local clock.




src/NTPClient_Generic_Impl.h
/****************************************************************************************************************************
  NTPClient_Generic_Impl.h
 
  For AVR, ESP8266/ESP32, SAMD21/SAMD51, nRF52, STM32, SAM DUE, WT32_ETH01, RTL8720DN, RP2040 boards using 
  1) Ethernet W5x00, ENC28J60, LAN8742A
  2) WiFiNINA
  3) ESP8266/ESP32 WiFi
  4) ESP8266/ESP32-AT-command WiFi
  5) WT32_ETH01 (ESP32 + LAN8720)
  6) RTL8720DN
  7) Portenta_H7
  8) RP2040W WiFi

  Based on and modified from Arduino NTPClient Library (https://github.com/arduino-libraries/NTPClient)
  to support other boards such as ESP8266/ESP32, SAMD21, SAMD51, Adafruit's nRF52 boards, SAM DUE, RTL8720DN, etc.
  and Ethernet/WiFi/WiFiNINA shields
  
  Copyright (C) 2015 by Fabrice Weinberg and licensed under MIT License (MIT)

  Built by Khoi Hoang https://github.com/khoih-prog/NTPClient_Generic
  Licensed under MIT license
  
  Version: 3.7.5

  Version Modified By  Date      Comments
  ------- -----------  ---------- -----------
  3.2.1   K Hoang      27/10/2020 Initial porting to support SAM DUE, SAMD21, SAMD51, nRF52, ESP32/ESP8266, STM32, etc. boards 
                                  using Ethernet/WiFi/WiFiNINA shields. Add more features and functions.
  3.2.2   K Hoang      28/10/2020 Add examples to use STM32 Built-In RTC.
  3.3.0   K Hoang      04/06/2021 Add support to RP2040-based boards. Add packet validity checks, version string and debug
  3.4.0   K Hoang      16/07/2021 Add support to WT32_ETH01 (ESP32 + LAN8720)
  3.5.0   K Hoang      10/08/2021 Add support to Ameba Realtek RTL8720DN, RTL8722DM and RTL8722CSM
  3.5.1   K Hoang      12/10/2021 Update `platform.ini` and `library.json`
  3.5.2   K Hoang      01/12/2021 Auto detect ESP32 core version. Fix bug in examples for WT32_ETH01
  3.6.0   K Hoang      08/01/2022 Fix `multiple-definitions` linker error. Add support to Portenta_H7
  3.7.0   K Hoang      20/01/2022 Make compatible to old code
  3.7.1   K Hoang      20/01/2022 Fix getUTCEpochMillis() bug
  3.7.2   K Hoang      23/02/2022 Add setUDP() function to enable auto-switching between WiFi and Ethernet UDP
  3.7.3   K Hoang      05/04/2022 Use Ethernet_Generic library as default. Support SPI1/SPI2 for RP2040/ESP32
  3.7.4   K Hoang      27/04/2022 Sync with NTPClient releases v3.2.1
  3.7.5   K Hoang      22/10/2022 Fix bug causing time jumping back or forth when network has problem
 *****************************************************************************************************************************/

#pragma once

// Reintroduce to prevent duplication compile error, pragma once can't prevent
#ifndef NTPCLIENT_GENERIC_IMPL_H
#define NTPCLIENT_GENERIC_IMPL_H

/////////////////////////////

NTPClient::NTPClient(const long& timeOffset)
{
  this->_timeOffset     = timeOffset;
}

/////////////////////////////

NTPClient::NTPClient(UDP& udp, const long& timeOffset)
{
  this->_udp            = &udp;
  this->_timeOffset     = timeOffset;
}

/////////////////////////////

NTPClient::NTPClient(UDP& udp, const char* poolServerName, const long& timeOffset, const unsigned long& updateInterval)
{
  this->_udp            = &udp;
  this->_timeOffset     = timeOffset;
  this->_poolServerName = poolServerName;
  this->_updateInterval = updateInterval;
}

/////////////////////////////

NTPClient::NTPClient(UDP& udp, const IPAddress& poolServerIP, const long& timeOffset, const unsigned long& updateInterval)
{
  this->_udp            = &udp;
  this->_timeOffset     = timeOffset;
  this->_poolServerIP   = poolServerIP;
  this->_poolServerName = NULL;
  this->_updateInterval = updateInterval;
}

/////////////////////////////

void NTPClient::begin(int port)
{
  this->_port = port;

  this->_udp->begin(this->_port);

  this->_udpSetup = true;
}

/////////////////////////////

// Perform some validity checks on the packet
//  https://datatracker.ietf.org/doc/html/rfc4330#section-4
// Check length before calling
static bool isValid(byte const *ntpPacket)
{
  unsigned long highWord = word(ntpPacket[16], ntpPacket[17]);
  unsigned long lowWord  = word(ntpPacket[18], ntpPacket[19]);

  unsigned long refTimeInt  = highWord << 16 | lowWord;

  highWord = word(ntpPacket[20], ntpPacket[21]);
  lowWord  = word(ntpPacket[22], ntpPacket[23]);

  unsigned long refTimeFrac = highWord << 16 | lowWord;

  byte leapIndicator = ((ntpPacket[0] & 0b11000000) >> 6);
  byte version       = ((ntpPacket[0] & 0b00111000) >> 3);
  byte stratum       =   ntpPacket[1];

  NTP_LOGDEBUG3("isValid: leapIndicator (!=3) =", leapIndicator, ", version (>=1) =", version);
  NTP_LOGDEBUG3("stratum (1 <= stratum <= 15) =", stratum, ", refTimeInt (!= 0) =", refTimeInt);
  NTP_LOGDEBUG1("refTimeFrac (!= 0) =", refTimeFrac);

  return
    (
      (leapIndicator !=  3) && // LI != UNSYNC
      (version       >=  1) &&
      (stratum       >=  1) &&
      (stratum       <= 15) &&
      ((refTimeInt != 0) || (refTimeFrac != 0))
    );
}

/////////////////////////////

bool NTPClient::checkResponse()
{
  if (this->_udp->parsePacket())
  {
    this->_lastRequest = 0; // no outstanding request
    int numBytesRead = this->_udp->read(this->_packetBuffer, NTP_PACKET_SIZE);

    NTP_LOGDEBUG1("numBytesRead (48) =", numBytesRead);

    if ((numBytesRead == NTP_PACKET_SIZE) && isValid(this->_packetBuffer))
    {
      this->_lastUpdate = millis();

      unsigned long highWord = word(this->_packetBuffer[40], this->_packetBuffer[41]);
      unsigned long lowWord  = word(this->_packetBuffer[42], this->_packetBuffer[43]);

      // combine the four bytes (two words) into a long integer
      // this is NTP time (seconds since Jan 1 1900):
      unsigned long secsSince1900 = highWord << 16 | lowWord;

      this->_currentEpoc = secsSince1900 - SEVENTYYEARS;

      highWord = word(this->_packetBuffer[44], this->_packetBuffer[45]);
      lowWord  = word(this->_packetBuffer[46], this->_packetBuffer[47]);
      this->_currentFraction = highWord << 16 | lowWord;

      // if the user has set a callback function for when the time is updated, call it
      if (_updateCallback)
      {
        _updateCallback(this);
      }

      return true;
    }
  }

  return false;
}

/////////////////////////////

bool NTPClient::forceUpdate()
{
  NTP_LOGDEBUG("forceUpdate from NTP Server");

  // flush any existing packets
  while (this->_udp->parsePacket() != 0)
    this->_udp->flush();

  this->sendNTPPacket();

#define DELAY_BETWEEN_CHECK_RESPONSE      10      // In ms

  for (uint16_t timeout = 0; timeout < TIMEOUT_WAITING_NTP_PACKET / DELAY_BETWEEN_CHECK_RESPONSE; timeout++)
  {
    delay ( DELAY_BETWEEN_CHECK_RESPONSE );

    if (this->checkResponse())
      return true;
  }

  return false;
}

/////////////////////////////

bool NTPClient::update()
{
  unsigned long now = millis();

  NTP_LOGDEBUG("Update from NTP Server");

  if ( ((_lastRequest == 0) && (_lastUpdate == 0)) ||                            // Never requested or updated
       ((_lastRequest == 0) && ((now - _lastUpdate) >= _updateInterval)) ||      // Update after _updateInterval
       ((_lastRequest != 0) && ((now - _lastRequest) > _retryInterval)) )        // Update if there was no response to the request
  {
    // setup the UDP client if needed
    if ( !this->_udpSetup || (this->_port != NTP_DEFAULT_LOCAL_PORT) )
    {
      this->begin();
    }

    return forceUpdate();
  }
  else
    return false;
}

/////////////////////////////

unsigned long long NTPClient::getUTCEpochMillis()
{
  unsigned long long epoch;

  epoch  = (unsigned long long)this->_currentEpoc * 1000;       // last time returned via server in millis
  epoch += this->_currentFraction / FRACTIONSPERMILLI;          // add the fraction from the server
  epoch += millis() - this->_lastUpdate;                        // add the millis that have passed since the last update

  return epoch;
}

/////////////////////////////

unsigned long long NTPClient::getEpochMillis()
{
  unsigned long long epoch;

  epoch = (this->_timeOffset * 1000) + getUTCEpochMillis();

  return epoch;
}

/////////////////////////////

String NTPClient::createFormattedTime(const unsigned long& rawTime) const
{
  unsigned long tempo   = hour(rawTime);
  String hoursStr       = tempo < 10 ? "0" + String(tempo) : String(tempo);

  tempo   = minute(rawTime);
  String minuteStr      = tempo < 10 ? "0" + String(tempo) : String(tempo);

  tempo   = second(rawTime);
  String secondStr      = tempo < 10 ? "0" + String(tempo) : String(tempo);

  return hoursStr + ":" + minuteStr + ":" + secondStr;
}

/////////////////////////////

String NTPClient::getFormattedTime() const
{
  return createFormattedTime(this->getEpochTime());
}

/////////////////////////////

String NTPClient::getFormattedUTCTime() const
{
  return createFormattedTime(this->getUTCEpochTime());
}

/////////////////////////////

String NTPClient::getFormattedDateTime() const
{
  char buf[32];
  char m[4];    // temporary storage for month string (DateStrings.cpp uses shared buffer)

  time_t t = this->getEpochTime();

  memset(buf, 0, sizeof(buf));
  memset(m, 0, sizeof(m));

  strcpy(m, monthShortStr(month(t)));
  sprintf(buf, "%2d:%2d:%2d %s %2d %s %d", hour(t), minute(t), second(t),
          dayShortStr(weekday(t)), day(t), m, year(t));

  return String(buf);
}

/////////////////////////////

String NTPClient::getFormattedUTCDateTime() const
{
  char buf[32];
  char m[4];    // temporary storage for month string (DateStrings.cpp uses shared buffer)

  time_t t = this->getEpochTime() - _timeOffset;

  memset(buf, 0, sizeof(buf));
  memset(m, 0, sizeof(m));

  strcpy(m, monthShortStr(month(t)));
  sprintf(buf, "%2d:%2d:%2d %s %2d %s %d", hour(t), minute(t), second(t),
          dayShortStr(weekday(t)), day(t), m, year(t));

  return String(buf);
}

/////////////////////////////

void NTPClient::sendNTPPacket()
{
  // set all bytes in the buffer to 0
  memset(this->_packetBuffer, 0, NTP_PACKET_SIZE);

  // Initialize values needed to form NTP request
  // (see URL above for details on the packets)
  this->_packetBuffer[0] = 0b11100011;   // LI, Version, Mode
  this->_packetBuffer[1] = 0;     // Stratum, or type of clock
  this->_packetBuffer[2] = 6;     // Polling Interval
  this->_packetBuffer[3] = 0xEC;  // Peer Clock Precision

  // 8 bytes of zero for Root Delay & Root Dispersion
  this->_packetBuffer[12]  = 49;
  this->_packetBuffer[13]  = 0x4E;
  this->_packetBuffer[14]  = 49;
  this->_packetBuffer[15]  = 52;

  // all NTP fields have been given values, now
  // you can send a packet requesting a timestamp:
  if  (this->_poolServerName)
  {
    this->_udp->beginPacket(this->_poolServerName, NTP_SEVER_PORT);
  }
  else
  {
    this->_udp->beginPacket(this->_poolServerIP, NTP_SEVER_PORT);
  }

  this->_udp->write(this->_packetBuffer, NTP_PACKET_SIZE);
  this->_udp->endPacket();

  this->_lastRequest = millis();
}

#endif    // NTPCLIENT_GENERIC_IMPL_H
